### ISSUE TITLE
OpenSSL/GnuTLS differences in cacert file handling

### ISSUE THREAD

**Issue by nbanb** [curl/curl#15544]
### I did this

HI 
Just after upgrading to curl 8.10.1 from debian repository (from curl 7.88.1), I have found that one of the main difference is that cURL 8 is builded using GNUTLS in replacement of OPENSSL

The trouble is that now cURL is not able to establish a secure HTTPS connection to a server when the server name (name of the URL) is a domain configured in certificate Subject Alternativ Name 

Here I will provide 2 request, the first request is made with cURL 7.88.1 using openssl 3+ and it success in verifying the certificate when the domain name we connect to is in the Subject Alternativ Names list of domain of  the  certificate.

The second request is exactly the same but with  curl 8.10.1 using GNUTLS and it failed to verify the certificate.

First request with curl7 and openssl :  result is OK

```bash
$ curl --version
curl 7.88.1 (x86_64-pc-linux-gnu) libcurl/7.88.1 OpenSSL/3.0.14 zlib/1.2.13 brotli/1.0.9 zstd/1.5.4 libidn2/2.3.3 libpsl/0.21.2 (+libidn2/2.3.3) libssh2/1.10.0 nghttp2/1.52.0 librtmp/2.3 OpenLDAP/2.5.13
Release-Date: 2023-02-20, security patched: 7.88.1-10+deb12u7
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd

$ curl -vvv --cacert /dev/shm/fbx-cacert https://fbx.********.net:20xx/
*   Trying xx.xx.xx.xx:20xx...
* Connected to fbx.********.net (xx.xx.xx.xx) port 20xx (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /dev/shm/fbx-cacert
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: C=FR; ST=PARIS; O=14RV; OU=14RV-NET; CN=*.14rv.lan; emailAddress=nba@14rv.lan
*  start date: Jan 31 10:52:03 2023 GMT
*  expire date: Feb  1 10:52:03 2025 GMT
*  subjectAltName: host "fbx.********.net" matched cert's "*.********.net"
*  issuer: C=FR; ST=PARIS; O=14RV; OU=14RV-NET; CN=14rv.lan; emailAddress=nba@14rv.lan
*  SSL certificate verify ok.
* using HTTP/1.1
> GET / HTTP/1.1
> Host: fbx.xxxxxxxx.net:20xx
> User-Agent: curl/7.88.1
> Accept: */*
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< HTTP/1.1 302 Found
< Server: nginx
< Date: Mon, 11 Nov 2024 09:19:52 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 0
< Connection: keep-alive
< Location: /login.php
< Expires: Mon, 11 Nov 2024 09:19:51 GMT
< Cache-Control: no-cache
< Cache-Control: must-revalidate,no-store
< 
* Connection #0 to host fbx.********.net left intact

```

Second request with curl 8 and gnutls :  result is KO

```bash
$ curl --version
curl 8.10.1 (x86_64-pc-linux-gnu) libcurl/8.10.1 GnuTLS/3.7.9 zlib/1.2.13 brotli/1.0.9 zstd/1.5.4 libidn2/2.3.3 libpsl/0.21.2 libssh2/1.10.0 nghttp2/1.52.0 ngtcp2/1.6.0 nghttp3/1.4.0 librtmp/2.3 OpenLDAP/2.5.13
Release-Date: 2024-09-18, security patched: 8.10.1-1~bpo12+1
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp ws wss
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd

$ curl -vvv --cacert /dev/shm/fbx-cacert https://fbx.********.net:20xx
10:24:22.637947 [0-x] == Info: [READ] client_reset, clear readers
10:24:22.687437 [0-0] == Info: Host fbx.******.net:20xx was resolved.
10:24:22.687683 [0-0] == Info: IPv6: 2a01:zzzz:zzzz:zzzz::1
10:24:22.687885 [0-0] == Info: IPv4: 8x.xx.xx.xx7
10:24:22.688069 [0-0] == Info: [HTTPS-CONNECT] added
10:24:22.688258 [0-0] == Info: [HTTPS-CONNECT] connect, init
10:24:22.688471 [0-0] == Info: [HTTPS-CONNECT] connect, check h21
10:24:22.688730 [0-0] == Info:   Trying [2a01:zzzz:zzzz:zzzz::1]:20xx...
10:24:22.689055 [0-0] == Info: [HTTPS-CONNECT] connect -> 0, done=0
10:24:22.689295 [0-0] == Info: [HTTPS-CONNECT] adjust_pollset -> 1 socks
10:24:22.689667 [0-0] == Info: [HTTPS-CONNECT] connect, check h21
10:24:22.689909 [0-0] == Info: [SSL] cf_connect()
10:24:22.690178 [0-0] == Info: GnuTLS ciphers: NORMAL:-ARCFOUR-128:-CTYPE-ALL:+CTYPE-X509:-VERS-SSL3.0
10:24:22.690592 [0-0] == Info: [SSL] No cached session ID for https://fbx.********.net:20xx
10:24:22.691282 [0-0] == Info: [SSL] handshake: outgoing message type 1
10:24:22.691588 [0-0] == Info: [SSL] gtls_push(len=405) -> 405, err=0
10:24:22.691834 [0-0] == Info: [SSL] loading trust anchors and CRLs
10:24:22.692911 [0-0] == Info: found 4 certificates in /dev/shm/fbx-cacert
10:24:22.755422 [0-0] == Info: found 447 certificates in /etc/ssl/certs
10:24:22.755544 [0-0] == Info: [SSL] glts_pull(len=5) -> -1, err=81
10:24:22.755692 [0-0] == Info: [SSL] cf_connect() -> 0, done=0
10:24:22.755814 [0-0] == Info: [HTTPS-CONNECT] connect -> 0, done=0
10:24:22.755967 [0-0] == Info: [SSL] adjust_pollset, POLLIN fd=5
10:24:22.756109 [0-0] == Info: [HTTPS-CONNECT] adjust_pollset -> 1 socks
10:24:22.756276 [0-0] == Info: [HTTPS-CONNECT] connect, check h21
10:24:22.756432 [0-0] == Info: [SSL] cf_connect()
10:24:22.756532 [0-0] == Info: [SSL] cf_connect() -> 0, done=0
10:24:22.756660 [0-0] == Info: [HTTPS-CONNECT] connect -> 0, done=0
10:24:22.756808 [0-0] == Info: [SSL] adjust_pollset, POLLIN fd=5
10:24:22.756945 [0-0] == Info: [HTTPS-CONNECT] adjust_pollset -> 1 socks
10:24:22.889293 [0-0] == Info: [HTTPS-CONNECT] connect, check h21
10:24:22.889516 [0-0] == Info: [SSL] cf_connect()
10:24:22.889700 [0-0] == Info: [SSL] cf_connect() -> 0, done=0
10:24:22.889913 [0-0] == Info: [HTTPS-CONNECT] connect -> 0, done=0
10:24:22.890144 [0-0] == Info: [SSL] adjust_pollset, POLLIN fd=5
10:24:22.890364 [0-0] == Info: [HTTPS-CONNECT] adjust_pollset -> 1 socks
10:24:22.927106 [0-0] == Info: [HTTPS-CONNECT] connect, check h21
10:24:22.927346 [0-0] == Info: [SSL] cf_connect()
10:24:22.927542 [0-0] == Info: [SSL] glts_pull(len=5) -> 5, err=0
10:24:22.927785 [0-0] == Info: [SSL] glts_pull(len=155) -> 155, err=0
10:24:22.928540 [0-0] == Info: [SSL] handshake: incoming message type 2
10:24:22.928841 [0-0] == Info: [SSL] gtls_push(len=6) -> 6, err=0
10:24:22.929162 [0-0] == Info: [SSL] glts_pull(len=5) -> 5, err=0
10:24:22.929392 [0-0] == Info: [SSL] glts_pull(len=1) -> 1, err=0
10:24:22.929622 [0-0] == Info: [SSL] glts_pull(len=5) -> 5, err=0
10:24:22.929860 [0-0] == Info: [SSL] glts_pull(len=58) -> 58, err=0
10:24:22.930095 [0-0] == Info: [SSL] handshake: incoming message type 8
10:24:22.930346 [0-0] == Info: [SSL] glts_pull(len=5) -> 5, err=0
10:24:22.930578 [0-0] == Info: [SSL] glts_pull(len=3252) -> 3252, err=0
10:24:22.930840 [0-0] == Info: [SSL] handshake: incoming message type 11
10:24:22.931089 [0-0] == Info: [SSL] glts_pull(len=5) -> 5, err=0
10:24:22.931318 [0-0] == Info: [SSL] glts_pull(len=1049) -> 1049, err=0
10:24:22.931583 [0-0] == Info: [SSL] handshake: incoming message type 15
10:24:22.933824 [0-0] == Info: [SSL] glts_pull(len=5) -> 5, err=0
10:24:22.934057 [0-0] == Info: [SSL] glts_pull(len=69) -> 69, err=0
10:24:22.934290 [0-0] == Info: [SSL] handshake: incoming message type 20
10:24:22.934588 [0-0] == Info: [SSL] handshake: outgoing message type 20
10:24:22.934872 [0-0] == Info: [SSL] gtls_push(len=74) -> 74, err=0
10:24:22.935290 [0-0] == Info: SSL connection using TLS1.3 / ECDHE_RSA_AES_256_GCM_SHA384
10:24:22.937802 [0-0] == Info: server verification failed: certificate error, no details available. (CAfile: /dev/shm/fbx-cacert CRLfile: none)
10:24:22.938322 [0-0] == Info: [SSL] cf_connect() -> 60, done=1
10:24:22.938571 [0-0] == Info: [HTTPS-CONNECT] connect, all failed
10:24:22.938821 [0-0] == Info: [HTTPS-CONNECT] connect -> 60, done=1
10:24:22.939085 [0-0] == Info: [WRITE] cw-out done
10:24:22.939277 [0-0] == Info: closing connection #0
10:24:22.939474 [0-0] == Info: [HTTPS-CONNECT] close
10:24:22.939671 [0-0] == Info: [SETUP] close
10:24:22.939845 [0-0] == Info: [SSL] close
10:24:22.941420 [0-0] == Info: [SSL] close
10:24:22.941532 [0-0] == Info: [SETUP] destroy
10:24:22.941642 [0-0] == Info: [HTTPS-CONNECT] destroy
curl: (60) server verification failed: certificate error, no details available. (CAfile: /dev/shm/fbx-cacert CRLfile: none)
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the webpage mentioned above.


``` 

Today I'm maintaining a lbrary (https://github.com/nbanb/fbx-delta-nba_bash_api.sh or https://github.com/freeboxos/freeboxos-bash-api) which nearly exclusively rely on cURL and I cannot switch it to insecure TLS (-k) in 2024 (maybe I could do that 10 years before, but definitly not in 2024 !)
I need for this project that cURL installed packages from current linux distributions repository  have the same behavior when compiled with openssl or with gnutls support (people must not have to build cURL manually to select openssl or gnutls ...)

Need to rollback to curl 7 for the moment 
Thanks for help
Kind regards
nbanba

  

### I expected the following

*  subjectAltName: host "fbx.********.net" matched cert's "*.********.net"
*  issuer: C=FR; ST=PARIS; O=14RV; OU=14RV-NET; CN=14rv.lan; emailAddress=nba@14rv.lan
*  SSL certificate verify ok.

### curl/libcurl version

curl 8.10.1

### operating system

Debian 12

**Comment by bagder** [comment]
Using an unknown CA cert file against an unknown server causes problems. But how are we supposed to figure out the error there are so many unknowns?

**Comment by nbanb** [comment]
Dear  @bagder 

Thanks for answer
To clarify, when you say : 

> Using an unknown CA cert file against an unknown server causes problems

Do you talk about a private CA which is issuing some certificate and one of the certificate is used by the remote server and the root CA public key is passed to curl : `curl --cacert /path/to/CA/pubkey $SERVER_URL`  ?  (note that the CA is properly added to `ca-certificate`  package system installed on debian 12)

The server is availiable through a public URL / IP. 
The CA and the certificate pubkey are 8192 bits 
If you need it, no problem but how can I send you (in private) the server URL and the rootCA pubkey?

Also, if you need some particular informations and / or me to do some tests, please ask (I would be happy to help but don't know how to).

Thanks and kind regards
nbanba

**Comment by nbanb** [comment]
Dear @bagder 
I sent you an email with more (private) details.
Please let me know how I can help.
Regards
nbanba

**Comment by nbanb** [comment]
Hi

Sorry to insist, it is not a problem of 'too many unknowns'  AND the regression is manifest !

1) Just build curl 8.11-DEV with GNUTLS (last version present in this repo) 
2) Just installed in parallel curl 7.88 builded with OPENSSL fropm debian12 repository

The regression is really manifest, please HELP !

![Capture d’écran du 2024-11-13 09-04-56](https://github.com/user-attachments/assets/dab3de76-099c-446b-aec0-1e7587f07c54)

@bagder Please notice that curl 7.88 do not think that 

> there are so many unknowns

to validate the remote certificate from the CA certificate bundle passed to curl with `--cacert /path/to/cacert-bundle` 

(and for me it's just a "standard operation" all TLS capable client must provided)

Kind regards
nbanba

**Comment by bagder** [comment]
> How can you say that there 'are too many unknowns' to make curl work with private CA ?

Because you did not provide enough information here for us to reproduce this problem.

(sure, you emailed details to me privately but I am quite busy for several weeks ahead so it might take some time before I can look at it)

**Comment by bagder** [comment]
> curl -s '' --cacert /path/to/cacert/bundle https://...

That's just an illegal command line that should never work. What did you expect to happen?

**Comment by nbanb** [comment]
Dear @bagder 

Sure it's an illegal form and it shouldn't work !  but that's not the case with curl <7.88.1 : 

```bash
$ curl --version
curl 7.88.1 (x86_64-pc-linux-gnu) libcurl/7.88.1 OpenSSL/3.0.11 zlib/1.2.13 brotli/1.0.9 zstd/1.5.4 libidn2/2.3.3 libpsl/0.21.2 (+libidn2/2.3.3) libssh2/1.10.0 nghttp2/1.52.0 librtmp/2.3 OpenLDAP/2.5.13
Release-Date: 2023-02-20, security patched: 7.88.1-10+deb12u8
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd

$ curl -s '' https://httpbin.org/ip 
{
  "origin": "123.45.67.89"
}
```
On contrary, curl 8+ seems to refuse this illegal form : 

```bash
$ src/curl --version
curl 8.11.1-DEV (x86_64-pc-linux-gnu) libcurl/8.11.1-DEV GnuTLS/3.8.8 zlib/1.2.13 brotli/1.0.9 zstd/1.5.4 libidn2/2.3.3 libpsl/0.21.2 libssh2/1.10.0 nghttp2/1.52.0 ngtcp2/1.2.0 nghttp3/1.1.0 librtmp/2.3 OpenLDAP/2.5.13
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp ws wss
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Largefile libz NTLM PSL SSL threadsafe TLS-SRP UnixSockets zstd

$ src/curl -s '' https://httpbin.org/ip 
curl: option : blank argument where content is expected
curl: try 'curl --help' or 'curl --manual' for more information
```

How did I detect that ? simply with a bad fullfilling of a BASH array : 

```bash
curl_test () {
local URL=$remote_target
local CURL_OPTIONS=("")
[[ -n "$_SESSION_TOKEN" ]] \
            && CURL_OPTIONS+=(-H "X-App-Auth: $_SESSION_TOKEN") \
            && CURL_OPTIONS+=(-X GET)
curl -s ${CURL_OPTIONS[@]} $URL
}
```
This function was working since years with curl <7.88.1 and now do not work with curl 8+.
Please note that it should never work (but it did with curl <7.88.1)

I explane :  if we decide to print all values in CURL_OPTIONS array (${CURL_OPTIONS[@]}), we can see that the function is starting to fullfill the array at the second position : 
```bash
${CURL_OPTIONS[0]} :  <empty>
${CURL_OPTIONS[1]} :  -H
${CURL_OPTIONS[2]} :  "X-App-Auth: $_SESSION_TOKEN"   
${CURL_OPTIONS[3]} :  -X   
${CURL_OPTIONS[4]} :  GET  
```

So when bash expand the array of curl option the result is : 

```bash
curl -s ${CURL_OPTIONS[@]} $URL
=
curl -s  ${CURL_OPTIONS[0]} ${CURL_OPTIONS[1]} ${CURL_OPTIONS[2]} ${CURL_OPTIONS[3]} ${CURL_OPTIONS[4]} $URL
=
curl -s   <empty> -H "X-App-Auth: $_SESSION_TOKEN"    -X  GET  $URL
=
curl -s  '' -H "X-App-Auth: $_SESSION_TOKEN"    -X  GET  $URL
```
And it was working fine before curl 8 (for me sure it's not normal that it had been working until curl 8)

PS : 
The right fullfilling of the array should be a function like : 

 
```bash
curl_test () {
local URL=$remote_target
local CURL_OPTIONS=("")
[[ -n "$_SESSION_TOKEN" ]] \
            && CURL_OPTIONS=(-H "X-App-Auth: $_SESSION_TOKEN") \
            && CURL_OPTIONS+=(-X GET)
curl -s ${CURL_OPTIONS[@]} $URL
}
```

Kind regards
nbanba

**Comment by bagder** [comment]
We now report error on this weird use. We did not before. End of story.

**Comment by nbanb** [comment]
Dear @icing 
Thanks for changing the issue title and thanks for time spent in troubleshooting this issue. 
I think your last answer will help in finding the root cause, I think I just touch something => if true I will go back here in a close future with détails.
Thanks again 
Kind regards 
nbanba

**Comment by nbanb** [comment]
Dear @icing 

Thanks a lot for answer and help.
For a better understanding of a and b CA certificates, both are used, a is an 8192bit ca cert signing 8192bit certs and b is a 4096bit ca cert signing 4096bit certs (2 disctinct root of the same PKI) and CA a is for external communication where CA b is for internal communication
From what you say, I did constat the same behavior: Openssl match the first CA certificate of the bundle with a matching CN which is not the case of gnutls (testing on a second URL which present a certificate signed by 'b' CA work with openssl if 'b' CA certificate is the first in the bundle - that's why the library which builded the CA certificate bundle always put the CA certificate of the target URL in first in the buldle file -) 

To test I did 2 bundle cacert with a and b, the  first in the order A+B and the second in the ordre B+A 

Testing with a bundle of a + b in this order with openssl and gnutls : 

![Capture d’écran du 2024-11-16 12-40-00](https://github.com/user-attachments/assets/0ec82171-335b-4cbc-9748-2acd7cd9ab17)
=> openssl validate but not gnutls

Testing with a bundle of b + a in this order with openssl and gnutls : 
![Capture d’écran du 2024-11-16 12-41-58](https://github.com/user-attachments/assets/56a37e96-0876-4fa9-87b3-c616c8fae787)
=> opensl failed but gnutls validate

So I constat that gnutls match the last cert of the bundle where openssl match the first cert of the bundle: **gnutls seems to read the bundle by the end** 

And you're right, it's not a curl issue and the issue is due to the different way openssl and gnutls backend read cacert bundle to validate remote server certificate.

I can now write a workarround depending of curl TLS backend , thanks ! => I think this issue can be closed.

Thanks again @icing and @bagder for help, really appreciate !
Kind regards
nbanba
